onnxruntime/core/optimizer/quick_gelu_fusion.cc

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/quick_gelu_fusion.h"
+
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/utils.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace onnxruntime::common;
+
+namespace onnxruntime {
+
+/**
+Rewrite x*sigmoid(alpha*x) or x*sigmoid(x) to QuickGelu.
+*/
+Status QuickGeluFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+  for (auto node_index : node_topology_list) {
+    auto* p_node = graph.GetNode(node_index);
+    if (!p_node) continue;
+
+    Node& node = *p_node;
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+
+    InlinedVector<std::reference_wrapper<Node>> nodes_to_fuse;
+
+    int alpha_index = -1;
+    float alpha = 1.0f;
+    if (graph_utils::IsSupportedOptypeVersionAndDomain(node, "Mul", {7, 13, 14}) &&
+        graph_utils::IsSupportedProvider(node, GetCompatibleExecutionProviders()) && node.GetOutputEdgesCount() == 1) {
+      for (auto i = 0; i < 2; ++i) {
+        const NodeArg& input_arg = *(node.InputDefs()[i]);
+        if (!optimizer_utils::IsScalar(input_arg)) continue;
+        const TensorProto* tensor_proto = graph_utils::GetConstantInitializer(graph, input_arg.Name());
+        if (!tensor_proto) continue;
+        Initializer init_const{*tensor_proto, graph.ModelPath()};
+        const auto data_type = tensor_proto->data_type();
+        if (data_type == TensorProto_DataType_FLOAT) {
+          alpha = *(init_const.data<float>());
+          alpha_index = i;
+          break;
+        } else if (data_type == TensorProto_DataType_DOUBLE) {
+          alpha = static_cast<float>(*(init_const.data<double>()));
+          alpha_index = i;
+          break;
+        } else if (data_type == TensorProto_DataType_FLOAT16) {
+          alpha = math::halfToFloat(init_const.data<MLFloat16>()->val);
+          alpha_index = i;
+          break;
+        }
+      }
+    }
+
+    NodeArg* quick_gelu_input_arg = nullptr;
+    Node* p_sigmoid_node = p_node;
+    if (alpha_index != -1) {
+      quick_gelu_input_arg = node.MutableInputDefs()[(alpha_index + 1) % 2];
+      nodes_to_fuse.emplace_back(node);
+      p_sigmoid_node = graph.GetNode(node.OutputNodesBegin()->Index());
```

Comment:
Can output of `Mul` be a graph output (node.OutputNodesBegin() is not indexable)?

Reply:
There is check for each Mul using node.GetOutputEdgesCount() == 1, which is for edge count, not for output def count. So there is one edge, and consumed by Sigmoid, so it's not possible to be graph output. The UT also has case to cover this.

Reply:
@iK1D @baijumeswani Would you mind taking a look at my PR #15109? My understanding is that even when Mul's `GetOutputEdgesCount()` is 1, the output of Mul can still be a graph output. The PR contains further elaboration on this point, along with two graph examples for unit testing purposes.